flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperatorStateHandler.java

```diff
@@ -0,0 +1,411 @@
+				"partitioned/keyed.");
+		}
+	}
+
+	@SuppressWarnings({"unchecked", "rawtypes"})
+	public void setCurrentKey(Object key) {
+		if (keyedStateBackend != null) {
+			try {
+				// need to work around type restrictions
+				@SuppressWarnings("unchecked,rawtypes")
+				AbstractKeyedStateBackend rawBackend = (AbstractKeyedStateBackend) keyedStateBackend;
+
+				rawBackend.setCurrentKey(key);
+			} catch (Exception e) {
+				throw new RuntimeException("Exception occurred while setting the current key context.", e);
+			}
+		}
+	}
+
+	@SuppressWarnings({"unchecked", "rawtypes"})
+	public Object getCurrentKey() {
+		if (keyedStateBackend != null) {
+			return keyedStateBackend.getCurrentKey();
+		} else {
+			throw new UnsupportedOperationException("Key can only be retrieved on KeyedStream.");
+		}
+	}
+
+	public Optional<KeyedStateStore> getKeyedStateStore() {
+		return Optional.ofNullable(keyedStateStore);
+	}
+
+	/**
+	 * Returns a {@link InternalTimerService} that can be used to query current processing time
+	 * and event time and to set timers. An operator can have several timer services, where
+	 * each has its own namespace serializer. Timer services are differentiated by the string
+	 * key that is given when requesting them, if you call this method with the same key
+	 * multiple times you will get the same timer service instance in subsequent requests.
+	 *
+	 * <p>Timers are always scoped to a key, the currently active key of a keyed stream operation.
+	 * When a timer fires, this key will also be set as the currently active key.
+	 *
+	 * <p>Each timer has attached metadata, the namespace. Different timer services
+	 * can have a different namespace type. If you don't need namespace differentiation you
+	 * can use {@link VoidNamespaceSerializer} as the namespace serializer.
+	 *
+	 * @param name The name of the requested timer service. If no service exists under the given
+	 *             name a new one will be created and returned.
+	 * @param namespaceSerializer {@code TypeSerializer} for the timer namespace.
+	 * @param triggerable The {@link Triggerable} that should be invoked when timers fire
+	 *
+	 * @param <N> The type of the timer namespace.
+	 */
+	public <K, N> InternalTimerService<N> getInternalTimerService(
+			String name,
+			TypeSerializer<N> namespaceSerializer,
+			Triggerable<K, N> triggerable) {
+
+		checkTimerServiceInitialization();
+
+		// the following casting is to overcome type restrictions.
+		KeyedStateBackend<K> keyedStateBackend = getKeyedStateBackend();
+		TypeSerializer<K> keySerializer = keyedStateBackend.getKeySerializer();
+		InternalTimeServiceManager<K> keyedTimeServiceHandler = (InternalTimeServiceManager<K>) timeServiceManager;
+		TimerSerializer<K, N> timerSerializer = new TimerSerializer<>(keySerializer, namespaceSerializer);
+		return keyedTimeServiceHandler.getInternalTimerService(name, timerSerializer, triggerable);
+	}
```

Comment:
Regarding having timer-related responsibility in state handler:
> Why not to have getInternalTimerService in the base class (or better outside) and pass stateHandler as a 
> parameter? In StreamOperatorStateHandler.snapshotState we can have a parameter similar to 
> snapshotStateAction for timers state.

(sorry for repeating this comment, but I couldn't find an answer)

Reply:
What base class do you have in mind? Do you mean to duplicate this code both in `AbstractStreamOperator` and `AbstractStreamOperatorV2`? Also what do you mean "outside"?

Reply:
I have moved the timers snapshotting logic into `InternalTimerSerivceManager` class